app/services/sync_service.rb

```diff
@@ -20,17 +36,43 @@ def sync(request_key, records, request_payload, report_errors_on_class: nil)
       if report_errors_on_class.present?
         uuid_field = "#{request_key.to_s.singularize}_uuid"
         success_ids.each do |id|
-          report_errors_on_class.find_by(uuid_field => id).update_columns(synced_at: Time.now, last_sync_errors: nil)
+          synced_record = report_errors_on_class.find_by(uuid_field => id)
+          sync_status[report_errors_on_class][:synced].add(synced_record.id)
         end
         JSON(response.body)['errors'].map do |error|
-          report_errors_on_class.find_by(uuid_field => error['id']).update_column(:last_sync_errors, error)
+          failed_record = report_errors_on_class.find_by(uuid_field => error['id'])
+          sync_status[report_errors_on_class][:failed_id_to_errors_hash][failed_record.id].to_a << error.except('id')
         end
       end
     rescue => error
       puts "Could not sync #{request_key}. Error: #{error.message}"
     end
   end
 
+  def mark_synced_and_failed
+    failed_patient_ids = sync_status[Patient][:failed_id_to_errors_hash].keys
+    failed_patient_ids.each do |id|
+      errors = sync_status[Patient][:failed_id_to_errors_hash][id]
+      Patient.find(id).update_columns(last_sync_errors: errors)
+    end
+
+    synced_patient_ids = sync_status[Patient][:synced_ids] - failed_patient_ids
+    synced_patient_ids.each do |id|
+      Patient.find(id).update_columns(synced_at: Time.now, last_sync_errors: nil)
+    end
+
+    failed_visit_ids = sync_status[Visit][:failed_id_to_errors_hash].keys
+    failed_visit_ids.each do |id|
+      errors = sync_status[Visit][:failed_id_to_errors_hash][id]
+      Visit.find(id).update_columns(last_sync_errors: errors)
+    end
+
+    synced_visit_ids = sync_status[Visit][:synced_ids] - failed_visit_ids
+    synced_visit_ids.each do |id|
+      Visit.find(id).update_columns(synced_at: Time.now, last_sync_errors: nil)
```

Comment:
@divs1210 This will fire N queries (1 for each `Visit`). What do you think of doing a bulk update instead?

```
Visit.where(id: synced_visit_ids).update_all(synced_at: Time.now, last_sync_errors: nil)
```

Note: `update_all` doesn't instantiate any object and hence doesn't run any callbacks, which is the expected behavior I guess.